src/index/src/bloom_filter/applier.rs

```diff
@@ -48,86 +31,121 @@ impl BloomFilterApplier {
                     }
-                }
-
-                is_any_range_hit |= matches;
-                if matches {
                     break;
                 }
             }
-            if !is_any_range_hit {
-                row_groups_to_remove.insert(row_group_idx);
-            }
         }
 
-        // 4. Remove row groups that do not match any bloom filter segment
-        for row_group_idx in row_groups_to_remove {
-            basement.remove(&row_group_idx);
+        Ok(ranges)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::atomic::AtomicUsize;
+    use std::sync::Arc;
+
+    use futures::io::Cursor;
+
+    use super::*;
+    use crate::bloom_filter::creator::BloomFilterCreator;
+    use crate::bloom_filter::reader::BloomFilterReaderImpl;
+    use crate::external_provider::MockExternalTempFileProvider;
+
+    #[tokio::test]
+    #[allow(clippy::single_range_in_vec_init)]
+    async fn test_appliter() {
+        let mut writer = Cursor::new(Vec::new());
+        let mut creator = BloomFilterCreator::new(
+            4,
+            Arc::new(MockExternalTempFileProvider::new()),
+            Arc::new(AtomicUsize::new(0)),
+            None,
+        );
+
+        let rows = vec![
+            vec![b"row00".to_vec(), b"seg00".to_vec(), b"overl".to_vec()],
+            vec![b"row01".to_vec(), b"seg00".to_vec(), b"overl".to_vec()],
+            vec![b"row02".to_vec(), b"seg00".to_vec(), b"overl".to_vec()],
+            vec![b"row03".to_vec(), b"seg00".to_vec(), b"overl".to_vec()],
+            vec![b"row04".to_vec(), b"seg01".to_vec(), b"overl".to_vec()],
+            vec![b"row05".to_vec(), b"seg01".to_vec(), b"overl".to_vec()],
+            vec![b"row06".to_vec(), b"seg01".to_vec(), b"overp".to_vec()],
+            vec![b"row07".to_vec(), b"seg01".to_vec(), b"overp".to_vec()],
+            vec![b"row08".to_vec(), b"seg02".to_vec(), b"overp".to_vec()],
+            vec![b"row09".to_vec(), b"seg02".to_vec(), b"overp".to_vec()],
+            vec![b"row10".to_vec(), b"seg02".to_vec(), b"overp".to_vec()],
+            vec![b"row11".to_vec(), b"seg02".to_vec(), b"overp".to_vec()],
+        ];
+
+        let cases = vec![
+            (vec![b"row00".to_vec()], 0..12, vec![0..4]), // search one row in full range
+            (vec![b"row05".to_vec()], 4..8, vec![4..8]),  // search one row in partial range
+            (vec![b"row03".to_vec()], 4..8, vec![]), // search for a row that doesn't exist in the partial range
+            (
+                vec![b"row01".to_vec(), b"row06".to_vec()],
+                0..12,
+                vec![0..8],
+            ), // search multiple rows in multiple ranges
+            (vec![b"row99".to_vec()], 0..12, vec![]), // search for a row that doesn't exist in the full range
+            (vec![b"row00".to_vec()], 12..12, vec![]), // search in an empty range
+            (
+                vec![b"row04".to_vec(), b"row05".to_vec()],
+                0..12,
+                vec![4..8],
```

Comment:
if search for "row00" and "row11" does it return `vec![0..12]` or `vec![0..4, 8..12]`?

Reply:
`vec![0..4, 8..12]`, since you are curious about this case, I have added it.